File: ts/src/samchon/protocol/slave/SlaveClient.ts
/// <reference path="../../API.ts" />

/// <reference path="SlaveSystem.ts" />

namespace samchon.protocol.slave
{
	export abstract class SlaveClient
		extends SlaveSystem
	{
		/* ---------------------------------------------------------
			CONSTRUCTORS
		--------------------------------------------------------- */
		/**
		 * Default Constructor.
		 */
		public constructor()
		{
			super();
		}

		/**
		 * @inheritdoc
		 */
		protected abstract createServerConnector(): IServerConnector;
		
		/* ---------------------------------------------------------
			METHOD OF CONNECTOR
		--------------------------------------------------------- */
		public connect(ip: string, port: number): void
		{
			if (this.communicator_ != null)
				return;

			this.communicator_ = this.createServerConnector();
			(this.communicator_ as IServerConnector).connect(ip, port);
		}
	}
}

File: cpp/samchon/templates/slave/SlaveClient.cpp
#include <samchon/templates/slave/SlaveClient.hpp>

#include <samchon/protocol/ServerConnector.hpp>

using namespace std;
using namespace samchon;
using namespace samchon::library;
using namespace samchon::protocol;
using namespace samchon::templates::slave;

/* ---------------------------------------------------------
	CONSTRUCTORS
--------------------------------------------------------- */
SlaveClient::SlaveClient()
	: SlaveSystem()
{
}
SlaveClient::~SlaveClient()
{
}

auto SlaveClient::createServerConnector() -> ServerConnector*
{
	return new ServerConnector(this);
}

/* ---------------------------------------------------------
	METHOD OF CONNECTOR
--------------------------------------------------------- */
void SlaveClient::connect(const string &ip, int port)
{
	shared_ptr<ServerConnector> connector(createServerConnector());
	this->communicator_ = connector;

	connector->connect(ip, port);
}